pyquil/api/_quantum_computer.py

```diff
@@ -669,45 +671,114 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
+
+    :returns: ``True`` if a process is listening on the specified host/port, ``False`` otherwise
+    """
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        s.connect((host, port))
+        return True
+    except ConnectionRefusedError:
+        return False
+    finally:
+        s.close()
+
+
+@contextmanager
+def local_forest_runtime(
+        *,
+        host: str = '127.0.0.1',
+        qvm_port: int = 5000,
+        quilc_port: int = 5555,
+        use_protoquil: bool = False) -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+    """A context manager for local QVM and QUIL compiler.
+
     You must first have installed the `qvm` and `quilc` executables from
     the forest SDK. [https://www.rigetti.com/forest]
 
-    This context manager will start up external processes for both the
-    compiler and virtual machine, and then terminate them when the context
-    is exited.
+    This context manager will ensure that the designated ports are not used, start up `qvm` and
+    `quilc` proccesses if possible and terminate them when the context is exited.
+    The returned tuple contains two ``subprocess.Popen`` objects for the `qvm` and the `quilc` processes.
+    If one of the ports is in use, a ``RuntimeWarning`` will be issued, the `qvm`/`quilc` process
+    won't be started and the respective value in the tuple will be ``None``.
+
+    .. note::
+        Only processes started by this context manager will be terminated on exit, no external process will
+        be touched.
 
-    If `qvm` (or `quilc`) is already running, then the existing process will
-    be used, and will not terminated at exit.
 
     >>> from pyquil import get_qc, Program
     >>> from pyquil.gates import CNOT, Z
-    >>> from pyquil.api import local_qvm
+    >>> from pyquil.api import local_forest_runtime
     >>>
     >>> qvm = get_qc('9q-square-qvm')
     >>> prog = Program(Z(0), CNOT(0, 1))
     >>>
-    >>> with local_qvm():
+    >>> with local_forest_runtime():
     >>>     results = qvm.run_and_measure(prog, trials=10)
 
+    :param host: Host on which `qvm` and `quilc` should listen on.
+    :param qvm_port: Port which should be used by `qvm`.
+    :param quilc_port: Port which should be used by `quilc`.
+    :param use_protoquil: Restrict input/output to protoquil.
+
+    .. warning::
+        If ``use_protoquil`` is set to ``True`` language features you need
+        may be disabled. Please use it with caution.
+
     :raises: FileNotFoundError: If either executable is not installed.
     """
-    # Enter. Acquire resource
-    qvm = subprocess.Popen(['qvm', '-S'],
-                           stdout=subprocess.PIPE,
-                           stderr=subprocess.PIPE)
 
-    quilc = subprocess.Popen(['quilc', '-RP'],
-                             stdout=subprocess.PIPE,
-                             stderr=subprocess.PIPE)
+    qvm = None
+    quilc = None
+
+    if _port_used('127.0.0.1', qvm_port):
```

Comment:
I think you should pass `host` here, rather than always passing `127.0.0.1`. Likewise for quilc below.

Reply:
I may be wrong, but the host we're accepting as a parameter is a server side argument, no?
E.g. someone may pass `0.0.0.0` to accept connections from all interfaces, in contrast to the host we pass to `_port_used`, which is the client-side parameter to specify which address to connect to.

Please correct me if I'm wrong, may be heading in the complete wrong direction here.

Reply:
You are correct, `host` here is the address the server will bind. The case of what to do with "0.0.0.0" is a good point. But if the caller specifies anything other than 0.0.0.0 or 127.0.0.1, then you don't necessarily want to query 127.0.0.1.

Maybe if they pass 0.0.0.0, check localhost; otherwise `host`?

Reply:
Also, for what it's worth, it common for implementations to allow you to `connect(2)` to 0.0.0.0, even though I think it's technically only valid as a source address?

Anyway, I just tested an Python on my macbook doesn't complain, and essentially treats it as if I had specified localhost, so maybe we don't have to do anything special... not sure how portable / reliable that is though.

Reply:
> Also, for what it's worth, it common for implementations to allow you to connect(2) to 0.0.0.0, even though I think it's technically only valid as a source address?

Interesting, didn't know that!

Just realised that it's possible to listen to the same port but bind to different addresses, so there may be a scenario where the user has an application e.g. listening to `127.0.0.1` on port 5000, but when binding to a different address, let's say `10.11.12.102`, we can still listen to port 5000 for qvm. (Summing up things to make sure I got it right)

Long story short, it completely makes sense to pass the host to `_check_port`. Happy to add an additional `if` to replace `0.0.0.0` with `127.0.0.` to be on the safe side?

Let me know!

Reply:
I think passing `host` and mapping 0.0.0.0 --> 127.0.0.1 is a reasonable compromise.

Even that is not technically correct, since it's not required that localhost is available or even that a loopback interface is configured, but it seems reasonable to assume 0.0.0.0 will bind to 127.0.0.1 in most environments were pyquil is running.

Reply:
Added!